lib/active_publisher/async/in_memory_adapter/async_queue.rb

```diff
@@ -60,11 +61,28 @@ def size
 
         def create_and_supervise_consumer!
           @consumer = ::ActivePublisher::Async::InMemoryAdapter::ConsumerThread.new(queue)
+          @last_heartbeat_at = ::Time.now
 
           supervisor_task = ::Concurrent::TimerTask.new(:execution_interval => supervisor_interval) do
-            unless consumer.alive?
+            current_time = ::Time.now
+
+            # Process heartbeats from the consumer.
+            if !consumer.heartbeats.empty?
+              @last_heartbeat_at = current_time
+              consumer.heartbeats.clear
+            end
+
+            # Consumer is lagging if we have not received a heartbeat from the consumer thread in more than
+            # 10 seconds.
+            seconds_since_last_heartbeat = current_time - @last_heartbeat_at
+            consumer_is_lagging = seconds_since_last_heartbeat > 10
+            logger.error "ActivePublisher consumer is lagging. Last heartbeat received #{seconds_since_last_heartbeat} seconds ago." if consumer_is_lagging
+
+            # Check to see if we should restart the consumer.
+            if !consumer.alive? || consumer_is_lagging
```

Comment:
when this is lagging, is it safe to kill the consumer?

Reply:
Depending on where the consumer thread is executing, it might drop the current batch of dequeued messages. This is true of the consumer 💥 ing on some other random error.

Reply:
wondering if we would lose any messages really